proposals/NNNN-resource-arrays.md

```diff
@@ -0,0 +1,258 @@
+`C[2][2][5]` as if it was `C[20]`.
+
+For example a call to initialize `C[1][0][3][0]` from the array declared above
+would have lower bound value `10`, upper bound `29`, space `1`, and array index
+`23` = `10 + 1*2*5 + 0*5 + 3`:
+```
+%dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 10, i32 29, i32 1, i8 1 }, i32 23, i1 false), 
+```
+
+https://godbolt.org/z/8MMWheeGc
+
+#### Arrays of Unbounded Size
+
+Groups of resource arrays can be declared with undefined size. These are called
+_unbounded arrays_:
+```
+RWBuffer<float> D[] : register(u5);
+```
+For unbounded arrays the `@dx.op.createHandleFromBinding` calls to initialize
+individual resource handles will have upper bound set to `-1`. For example an
+initialization of handle `D[100]` from the `D` array declared above looks like
+this:
+```
+%dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 5, i32 -1, i32 0, i8 1 }, i32 105, i1 false)
+```
+https://godbolt.org/z/hx99s5dzP
+
+#### Unbounded-size Arrays and Multiple Dimensions
+
+For multi-dimensional arrays of resources only the first dimension can be
+unbounded:
+```
+RWBuffer<float> E[][4] : register(u5);
+```
+When initializing individual resources of an unbounded multi-dimensional array,
+the arguments of `@dx.op.createHandleFromBinding` again look like as if the
+array was flattened to a single dimension. For example the call to create
+resource handle `E[100][4]` from array `E` declared above will have index `409`
+= `5 + 4 x 100 + 4`:
+```
+call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 5, i32 -1, i32 0, i8 1 }, i32 409, i1 false),
+```
+https://godbolt.org/z/YrEnYhrEj
+
+Arrays with unbounded sizes in the other dimensions are not allowed, though the
+error message DXC reports does not clearly indicate this:
+
+```
+RWBuffer<float> F[4][];
+
+error: array has incomplete element type 'RWBuffer<float> []'
+```
+https://godbolt.org/z/osq7dzvdh
+
+### Resource Arrays at Local Scope
+
+#### Local Resource Arrays with Fixed Size
+
+Resources array with fixed size can be declared as local variables in a function.
+Resource handles for the individual resources in the array are uninitialized
+until they are assigned a value from an initialized resource. DXC does not
+report error when an uninitialized resource is accessed and instead creates a
+call to `dx.op.annotateHandle` with zero-initialized handle value.
+
+```
+RWBuffer<float> Buf;
+RWBuffer<float> Out;
+
+float foo() {
+  RWBuffer<float> LocalArray[3];
+  LocalArray[2] = Buf;
+  // LocalArray[0] is not initialized
+  return LocalArray[2][0] + LocalArray[0][0];
+}
+
+[numthreads(4,1,1)]
+void main() {
+  Out[0] = foo();
+}
+```
+https://godbolt.org/z/Gdc3q674W
+
+#### Unbounded Local Resource Arrays
+
+DXC even allows unbounded resource arrays as local variables in a function. This
+seems wrong from a language point of view. How is the compiler supposed to
+allocate such variable? 
```

Comment:
I strongly believe this should be considered a bug that DXC allows this.

Reply:
I agree.